From 7813af8415a44055f5f86716564d77b2d7f8a315 Mon Sep 17 00:00:00 2001
From: Frank Hunleth <fhunleth@troodon-software.com>
Date: Mon, 8 Jul 2019 10:44:12 -0400
Subject: [PATCH] rpi-firmware: support custom version selection

Since the rpi-firmware supplies firmware binaries that have the
potential to depend on kernel drivers, we've matched kernel versions to
rpi-firmware versions. In order to simplify using different Linux kernel
versions, this adds an option to set the tag or hash to use. For
example, if you're using the Raspberry Pi Linux kernel tagged
`1.20200902`, you can now use the rpi-firmware version that was tagged
`1.20200902`.
---
 package/rpi-firmware/Config.in       |  5 +++++
 package/rpi-firmware/rpi-firmware.mk | 14 +++++++++++++-
 2 files changed, 18 insertions(+), 1 deletion(-)

diff --git a/package/rpi-firmware/Config.in b/package/rpi-firmware/Config.in
index f99c116ca2..9b56954f5c 100644
--- a/package/rpi-firmware/Config.in
+++ b/package/rpi-firmware/Config.in
@@ -11,6 +11,11 @@ config BR2_PACKAGE_RPI_FIRMWARE
 
 if BR2_PACKAGE_RPI_FIRMWARE
 
+config BR2_PACKAGE_RPI_FIRMWARE_CUSTOM_VERSION
+	string "custom version"
+	help
+	  If specified, use this tag or git hash instead of the default
+
 config BR2_PACKAGE_RPI_FIRMWARE_BOOTCODE_BIN
 	bool "rpi 0/1/2/3 bootcode.bin"
 	help
diff --git a/package/rpi-firmware/rpi-firmware.mk b/package/rpi-firmware/rpi-firmware.mk
index 7ae9ff6c28..0d8a2afa92 100644
--- a/package/rpi-firmware/rpi-firmware.mk
+++ b/package/rpi-firmware/rpi-firmware.mk
@@ -4,7 +4,19 @@
 #
 ################################################################################
 
-RPI_FIRMWARE_VERSION = 5476720d52cf579dc1627715262b30ba1242525e
+ifeq ($(call qstrip,$(BR2_PACKAGE_RPI_FIRMWARE_CUSTOM_VERSION)),)
+ifeq ($(BR2_PACKAGE_RPI_FIRMWARE),y)
+$(error The Raspberry Pi firmware version is no longer derived automatically.\
+    Please set BR2_PACKAGE_RPI_FIRMWARE_CUSTOM_VERSION in your nerves_defconfig\
+    to a version that is compatible with the Linux kernel that you are using.\
+    For example, if using a tagged Raspberry Pi kernel, there usually is a \
+    corresponding tag at https://github.com/raspberrypi/firmware)
+endif
+else
+RPI_FIRMWARE_VERSION = $(call qstrip,$(BR2_PACKAGE_RPI_FIRMWARE_CUSTOM_VERSION))
+BR_NO_CHECK_HASH_FOR += $(RPI_FIRMWARE_SOURCE)
+endif
+
 RPI_FIRMWARE_SITE = $(call github,raspberrypi,firmware,$(RPI_FIRMWARE_VERSION))
 RPI_FIRMWARE_LICENSE = BSD-3-Clause
 RPI_FIRMWARE_LICENSE_FILES = boot/LICENCE.broadcom
-- 
2.43.0

